core/src/main/java/hudson/scheduler/BaseParser.java

```diff
@@ -70,15 +70,23 @@ public void setHash(Hash hash) {
     protected long doRange(int start, int end, int step, int field) throws ANTLRException {
         rangeCheck(start, field);
         rangeCheck(end, field);
+        if (field == 1 && start>end) // Hour-value passes midnight
+            end += 24;
         if (step <= 0)
             error(Messages.BaseParser_MustBePositive(step));
         if (start>end)
             error(Messages.BaseParser_StartEndReversed(end,start));
 
         long bits = 0;
-        for (int i = start; i <= end; i += step) {
-            bits |= 1L << i;
-        }
+        if (field == 1)
+            for (int i = start; i <= end; i += step) {
+                // Make sure the hours past midnight are smaller than 24
+                bits |= 1L << i % 24;
```

Comment:
What happens with `23-0`?

If the previous comment is correct, what about `21-60`?

Reply:
```23-0``` will get added as if they were defined as ```0,23```.
```21-60``` is not possible, as stated above.